packages/mermaid/src/diagrams/pie/pieDb.ts

```diff
@@ -0,0 +1,99 @@
+import { log } from '../../logger.js';
+import { parseDirective as _parseDirective } from '../../directiveUtils.js';
+import { getConfig as commonGetConfig } from '../../config.js';
+import { sanitizeText } from '../common/common.js';
+import {
+  setAccTitle,
+  getAccTitle,
+  setDiagramTitle,
+  getDiagramTitle,
+  getAccDescription,
+  setAccDescription,
+  clear as commonClear,
+} from '../../commonDb.js';
+import type { ParseDirectiveDefinition } from '../../diagram-api/types.js';
+import type { PieFields, PieDB, Sections } from './pieTypes.js';
+import type { RequiredDeep } from 'type-fest';
+import type { PieDiagramConfig } from '../../config.type.js';
+import { cleanAndMerge } from '../../utils.js';
+
+export const DEFAULT_PIE_CONFIG: Required<PieDiagramConfig> = {
+  useMaxWidth: true,
+  useWidth: 984,
+  textPosition: 0.75,
+} as const;
```

Comment:
Ideally, we should be loading these values from the JSON Schema, if they exist.

That way, all of these default values with be visible in the docs, like https://mermaid.js.org/config/schema-docs/config.html#definitions-group-piediagramconfig

```suggestion
import DEFAULT_CONFIG from '../../defaultConfig.js';

export const DEFAULT_PIE_CONFIG: Required<PieDiagramConfig> = DEFAULT_CONFIG.pie;
```

Reply:
I agree with you that we need load values from JSON Schema, but it's not fully typed and what to do with values that they're `undefined`? Using `Required` here would throw errors.

And we need to change the default config to `RequiredDeep<MermaidConfig>`.

Reply:
> but it's not fully typed and what to do with values that they're `undefined`? Using Required here would throw errors.

You're right. Ideally we also need to fix the `defaulConfig.ts` types first, but that'd need to wait until Mermaid v11 (see https://github.com/orgs/mermaid-js/discussions/4710#discussioncomment-6709156).

Is it worth doing something like the following instead? Maybe it's overkill since we'll remove most of it once we get the types working, though :shrug:. Up to you.

```typescript
import DEFAULT_CONFIG from '../../defaultConfig.js';

function isValidDefaultPieConfig(x: PieDiagramConfig): x is  Required<PieDiagramConfig> {
  return x.useMaxWidth && x.useWidth && x.textPosition;
};

export const DEFAULT_PIE_CONFIG: Required<PieDiagramConfig> = (() => {
  const defaultConfig = {useWidth: 984, ...DEFAULT_CONFIG.pie};
  if (!isValidDefaultPieConfig(defaultConfig)) {
    throw new Error("DEFAULT_PIE_CONFIG is missing required entries: ", defaultConfig);
  }
  return defaultConfig;
});
```

> And we need to change the default config to `RequiredDeep<MermaidConfig>`.

This works for Pie diagrams, but it won't work for all diagram configs, because some values actually can be `undefined` and are `undefined` be default, see https://github.com/mermaid-js/mermaid/pull/4501#discussion_r1292335636.

Reply:
> Is it worth doing something like the following instead? Maybe it's overkill since we'll remove most of it once we get the types working, though shrug. Up to you.

I think it's kinda overengineering to do so.

Reply:
> _**Note**_: you also flipped default values of `useWidth` and `useMaxWidth` in example :)

---

> ```js
>   pie: {
>     useWidth: 984, // should this be 1200 instead?
>   },
> ```

As I said before, `useWidth` is kinda vogue, the width in pie depends on the parent element, e.g., in `cypress/integration/rendering/pie.spec.ts` the width is `984`:

<https://github.com/mermaid-js/mermaid/blob/3d117815425cbd05105ec0f0192adc0be91931f2/cypress/integration/rendering/pie.spec.js#L54-L57>

But in `demos/pie.html`, the parent element width here is `1904`.

But the previous (apparently) default value is `1200`:

<https://github.com/mermaid-js/mermaid/blob/3d117815425cbd05105ec0f0192adc0be91931f2/packages/mermaid/src/diagrams/pie/pieRenderer.js#L42-L44>

It was discussed [here](https://github.com/mermaid-js/mermaid/pull/4501#discussion_r1235139653) and [here](https://github.com/mermaid-js/mermaid/pull/4501#discussion_r1291228283).

---

> Then, in this file, we just tell TypeScript this is a `Required<PieDiagramConfig>` object, e.g. something like:
> 
> ```js
> // @ts-expect-error: This is valid, since the `defaultConfig` defined in packages/mermaid/src/defaultConfig.ts has no `undefined` values
> const getConfig = (): Required<PieDiagramConfig> => commonGetConfig().pie;
> ```
> 
> It's not perfect, but you're right, it's probably over-engineering to do anything more.

Thanks! I'll do as you suggested.

Reply:
But I'm not sure why we're returning config from current global config.

Reply:
Since the schema import expect errors:

<https://github.com/mermaid-js/mermaid/blob/defe40692ae3b9646d4faa4495694acf5dc7e144/packages/mermaid/src/defaultConfig.ts#L8-L9>

We can change the `defaultConfig` type into `RequiredDeep` without errors (although other values are actually `undefined`) defe40692

Reply:
> But I'm not sure why we're returning config from current global config.

In the `packages/mermaid/src/config.ts` file, the `getSiteConfig()` returns:
  - The defaultConfig, merged with
  - the global site config (e.g. set by `mermaid.initialize()`)

In the `packages/mermaid/src/config.ts` file, the `getConfig()` returns the diagram specific config, e.g.:
  - The defaultConfig, merged with
  - the global site config (e.g. set by `mermaid.initialize()`)
  - **the diagram specific config (e.g. set by directives like `%%{init: {"flowchart": {"defaultRenderer": "elk"}} }%%`**

I know it's a bit confusing, but even though `getConfig()` in `packages/mermaid/src/config.ts` is globally available, because it's reset in-between diagram renders, and Mermaid renders each diagram sequentially, it actually contains the diagram-specific config that is set in `%%{init:` directives (and maybe YAML directives once we get it set up).

It's definitely worth changing in a future PR, though, since I know the behavior is pretty weird, and it causes bugs like https://github.com/mermaid-js/mermaid/issues/4345